### PR TITLE
[WIP]: Added Matrix groups

### PR DESCRIPTION
**Pull request by deepakgouda**
Matrix Groups were added in _sympy/combinatorics/matrix_groups.py_

Added Special Linear Groups, Special Orthogonal Groups(2x2 only) and matrix groups with entries as integer modulo p.

To Do:
- [ ] Add maximum and minimum values option for Special Linear Groups.

### CODE REVIEW

**Comment by Roland-Sands** [comment]
I think you should create a groups folder, and put matrix groups in it
Then I think you should create a folder called algebra, and put it in sympy/sympy
Then I think you should put the groups folder, the algebras folder and the liealgebras folder in the algebra folder
I also think that free groups and some of the other group stuff in combinatorics should be moved to the groups folder, but that might be harder

**Comment by Roland-Sands** [comment]
I think it would be helpful if you stated what you are trying to achieve
If it is along the lines of "Implement basic functionality of matrix groups" then I think you should write something like that

**Comment by Roland-Sands** [comment]
> I think you should create a groups folder, and put matrix groups in it
Then I think you should create a folder called algebra, and put it in sympy/sympy
Then I think you should put the groups folder, the algebras folder and the liealgebras folder in the algebra folder
I also think that free groups and some of the other group stuff in combinatorics should be moved to the groups folder, but that might be harder

I think you should ignore this
I think you should create a groups folder, put matrix groups in it, and then put the groups folder in sympy/sympy

**Comment by Roland-Sands** [comment]
On Gitter you wrote "Please review"
I am happy to give my opinion, but I am not an official person
My opinion is irrelevant to whether or not your code is merged

**Comment by deepakgouda** [comment]
@Roland-Sands The way I am generating the Special Linear Group is, I am generating two matrices, one is a upper triangular and the other is a lower triangular with diagonal entries as 1. Then I am multiplying them to get the matrix. Is there a better way to do it?

And what other functionalities shall we add? Shall we implement a method to check if a matrix is Special Linear Group? And a matrix with entries [[cos(x), sin(x)], [-sin(x), cos(x)]] has been done. Can we raise it to higher orders?

**Comment by deepakgouda** [comment]
I shall ping others to check the coding details. @gxyd @valglad

**Comment by valglad** [comment]
There have been suggestions of moving all group theory stuff to a separate module for some time but it hasn't happened yet. I'd ask again on gitter about this (maybe in the [Group Theory chat](https://gitter.im/sympy/GroupTheory)), and it could be cleaner to have a separate pull request that creates a group theory folder and moves everything there.

For now, I'd suggest you just stick with working in the combinatorics module. I'll have a look at this PR tomorrow (hopefully), if someone doesn't before then.

**Comment by Roland-Sands** [comment]
> The way I am generating the Special Linear Group is, I am generating two matrices, one is a upper triangular and the other is a lower triangular with diagonal entries as 1. Then I am multiplying them to get the matrix. Is there a better way to do it

`S` and `T` are a generating set for `SL_2(Z)`, where `S = Matrix(2,2, [0, -1, 1, 0])` and `T = Matrix(2,2, [1, 1, 0, 1])` (reference: http://www.math.uconn.edu/~kconrad/blurbs/grouptheory/SL(2,Z).pdf)
Also in the reference is an algorithm to write an element of `SL_2(Z)` in terms of these generators

http://www.math.uconn.edu/~kconrad/blurbs/grouptheory/SL(2,R).pdf has an algorithm for the Iwasawa decomposition of an element of `SL_2(R)`, I think it would be fantastic if that was implemented

> And what other functionalities shall we add? Shall we implement a method to check if a matrix is Special Linear Group? And a matrix with entries [[cos(x), sin(x)], [-sin(x), cos(x)]] has been done. Can we raise it to higher orders?

I will need a bit of time for this

**Comment by Roland-Sands** [comment]
If the Iwasawa decomposition for `SL(n,R)` was implemented then I think that would be sensational

**Comment by Roland-Sands** [comment]
> There have been suggestions of moving all group theory stuff to a separate module for some time but it hasn't happened yet

Free groups may have a combinatorial vibe, but it is unequivocally wrong to have `SL_2` in combinatorics

File: sympy/printing/latex.py [hunk 1]
```diff
@@ -780,12 +780,14 @@ def _print_UndefinedFunction(self, expr):
     @property
     def _special_function_classes(self):
         from sympy.functions.special.tensor_functions import KroneckerDelta
-        from sympy.functions.special.gamma_functions import gamma, lowergamma
+        from sympy.functions.special.gamma_functions import gamma,\
```
Comment (valglad) [inline sympy/printing/latex.py]: The modifications in this file don't seem to be related to linear groups. Are then in this PR by accident?

File: sympy/combinatorics/matrix_groups.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
+from __future__ import print_function
+from sympy.core import Basic
+from math import sin, cos, pi, floor
+import random
+
+def MatrixMultiply(a, b):
```
Comment (valglad) [inline sympy/combinatorics/matrix_groups.py]: SymPy already can multiply matrices simply by `a*b`.

File: sympy/combinatorics/matrix_groups.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    _arr = [[] for i in range(len(a))]
+    for i in range(len(a)):
+        for j in range(len(b[0])):
+            c = 0
+            for k in range(len(b)):
+                c = c + (a[i][k] * b[k][j])
+            _arr[i].append(c)
+    return _arr
+
+def isPrime(p):
```
Comment (valglad) [inline sympy/combinatorics/matrix_groups.py]: This already exists in `sympy.ntheory.primetest` and it's called `isprime`. See http://docs.sympy.org/latest/modules/ntheory.html#sympy.ntheory.primetest.isprime

File: sympy/combinatorics/matrix_groups.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        raise ValueError("p must be a Prime Number")
+
+    _finite_matrix_group = FiniteMatrixGroup(dim, p)
+    return _finite_matrix_group
+
+
+class SplLinearGroup(Basic):
+    def __new__(cls, dim, _min, _max):
+        """
+        Generates two dim x dim matrices, one upper triangular
```
Comment (valglad) [inline sympy/combinatorics/matrix_groups.py]: A single matrix is not a group so it doesn't make sense to call this class SplLinearGroup. Really, what you are doing here is returning a somewhat random invertible matrix over the integers. This is not really adding anything particularly new to what's already present in the matrices module. Have a look at the documentation for matrices: http://docs.sympy.org/latest/modules/matrices/index.html

File: sympy/combinatorics/matrix_groups.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        and other lower triangular with diagonal entries as 1
+        and other entries as random integers. The resultant of
+        their multiplication is returned.
+        """
+        random.seed()
+        a = [[] for i in range(dim)]
```
Comment (valglad) [inline sympy/combinatorics/matrix_groups.py]: You should be using instances of `Matrix` from the matrices module.

File: sympy/combinatorics/matrix_groups.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    def __new__(cls, dim, x = None):
+        _arr = [[] for i in range(dim)]
+        if x is None:
+            random.seed()
+            x = random.uniform(0, 2*pi)
+
+        _arr[0].append(cos(x))
+        _arr[0].append(sin(x))
+        _arr[1].append(-sin(x))
+        _arr[1].append(cos(x))
```
Comment (valglad) [inline sympy/combinatorics/matrix_groups.py]: If `dim>2`, you are only filling in the top 4 corner entries and leaving the rest of the entries undefined. Besides, for greater dimensions, this is not what a member of the orthogonal group would look like in general.

**Comment by valglad** [comment]
Overall, I don't think you thought this through. I'd suggest discussing exactly what you'd like to implement on gitter. If we were to have a class for linear groups, what attributes and methods would we want that class to have? What class would the elements of the group belong to (e.g. would they be instances of `Matrix` or maybe a subclass `MatrixGroupElement` inheriting from `Matrix` or something like that)? There are a few things to think about. Once this is all carefully planned, you could give it another go but for now I would suggest you close this PR.

File: sympy/combinatorics/matrix_groups.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        raise ValueError("Size must be greater than or equal to 1")
+    elif dim is 1:
+        return [1]
+
+    _spl_linear_group = SplLinearGroup(dim, min_val, max_val)
+    return _spl_linear_group
+
+def spl_orthogonal_group(dim = 2, x = None):
+    """
+    Returns a Special Orthogonal Group of 2x2 matrices with entries as
```
[the remark this answers is not in the document]
Comment (deepakgouda) [reply sympy/combinatorics/matrix_groups.py]: I intended to implement it for higher dimensions in the final code. But I wasn't able to do it now, so I implemented it for 2x2 and had asked for some suggestions.

**Comment by deepakgouda** [comment]
@valglad As I explored more of the libraries of sympy, I realized that this PR needs a lot of improvement and it wasn't carried out properly. So, I didn't do any commits to this PR. I have been working on and will make another PR soon(I've been busy for my exams) and close this PR. Thank you for the suggestions and I implement them soon.
